test/e2e/network/service.go

```diff
@@ -1971,6 +1972,88 @@ var _ = SIGDescribe("Services", func() {
 		framework.ExpectNoError(err, "failed to create service: %s in namespace: %s", serviceName, ns)
 	})
 
+	ginkgo.It("should not allocate NodePorts for Type LoadBalancer if allocateLoadBalancerNodePorts=false [Feature:ServiceLBNodePortControl]", func() {
+		serviceName := "lb-disable-nodeport"
+		ns := f.Namespace.Name
+
+		t := NewServerTest(cs, ns, serviceName)
+		defer func() {
+			defer ginkgo.GinkgoRecover()
+			errs := t.Cleanup()
+			if len(errs) != 0 {
+				framework.Failf("errors in cleanup: %v", errs)
+			}
+		}()
+
+		service := t.BuildServiceSpec()
+		service.Spec.Type = v1.ServiceTypeLoadBalancer
+		service.Spec.AllocateLoadBalancerNodePorts = utilpointer.BoolPtr(false)
+
+		ginkgo.By("creating service " + serviceName + " with type LoadBalancer and no node ports in namespace " + ns)
+		service, err := t.CreateService(service)
+		framework.ExpectNoError(err, "failed to create service: %s in namespace: %s", serviceName, ns)
+
+		foundNodePorts := false
+		for _, port := range service.Spec.Ports {
+			if port.NodePort > 0 {
+				foundNodePorts = true
+			}
+		}
+
+		if foundNodePorts {
+			framework.Failf("service %s/%s had unexpected node ports allocated", ns, serviceName)
+		}
+	})
+
+	ginkgo.It("should not de-allocate NodePorts for LoadBalancer if allocateLoadBalancerNodePorts switches from true to false [Feature:ServiceLBNodePortControl]", func() {
+		serviceName := "lb-disable-nodeport"
+		ns := f.Namespace.Name
+
+		t := NewServerTest(cs, ns, serviceName)
+		defer func() {
+			defer ginkgo.GinkgoRecover()
+			errs := t.Cleanup()
+			if len(errs) != 0 {
+				framework.Failf("errors in cleanup: %v", errs)
+			}
+		}()
+
+		service := t.BuildServiceSpec()
+		service.Spec.Type = v1.ServiceTypeLoadBalancer
+		service.Spec.AllocateLoadBalancerNodePorts = utilpointer.BoolPtr(true)
+
+		ginkgo.By("creating service " + serviceName + " with type LoadBalancer and with node ports in namespace " + ns)
+		service, err := t.CreateService(service)
+		framework.ExpectNoError(err, "failed to create service: %s/%s", ns, serviceName)
+
+		foundNodePorts := false
+		for _, port := range service.Spec.Ports {
+			if port.NodePort > 0 {
+				foundNodePorts = true
+			}
+		}
```

Comment:
For e2e we should poll the nodePort and verify here that there is connectivity, use the echo handler in agnhost per example.
Once we deallocate the nodePorts, and check that  the API has removed them from the spec, ,we should verify that the nodePort is not working.

Reply:
I don't really think we need to validate node ports data path in this test, we have plenty of other tests validating that behavior. Just checking that the node port was allocated in an integration test should be sufficient.

Reply:
is not testing the nodeport itself, it's testing that the CNI implementation really stops to forward the traffic there ;)
I really think we should do it, if we have the API fields the CNIs have to implement it, if you don't check the connectivity the test will pass but the CNI will not support it ... at the end of the day the e2e has to be a validation of all the APIs

Reply:
This is node ports so not sure how CNI is involved, can you expand on that?

Reply:
sorry, maybe I'm getting the feature wrong, but this is for a ServiceType loadbalancer right?
so, replace my bad wording, instead of CNI, for "whatever provides connectivity in the cluster to the Services", it may happen that it doesn't take into account NodePorts removal because of the ServiceType, so it will keep forwarding the traffic ... if it doesn't implement this feature it can assume LoadBalancer always have nodePorts, it's rare but seems feasible

Reply:
I think it makes sense if we can create a Service Type=LoadBalancer using direct pod-routing. I believe there's an annotation for the GCP provider (or similar) that can enable this. If the load balancer implementation itself is nodeport based then I see little value in that.

Reply:
ahh, nevermind then ... forget about my comment, maybe too paranoid :upside_down_face:

Reply:
... then all these tests are going to be API only